Shell script

chinarjoshi opened the pull request:
Overrides command not found handler and implements 2 keybindings:
  1. **Control + Space**: Toggle the buffer between natural language and bash.
  2. **Enter**: If command not found, then if the command is natural language, translate and execute.
  3. **Control + Enter**: Execute the natural language in the buffer, regardless whether its also a valid bash command *("find all files longer than 20 lines")*. Also useful for Mac because of the uppercase commands.

One last feature to implement is more robust checking for destructive commands.

chinarjoshi commented:
The shell script is now complete. Can someone please test it and check for any fixes?

chinarjoshi commented:
Something else to consider is putting this shell script in the main branch as a hidden file ".sgpt.sh", and on installation prompt them to opt in or out of modifying their shell rc to source it. I think that would allow more users to discover the feature compared to an install script from the README.

TheR1D reviewed (commented):
Thank you for the PR. I've noticed a few things during testing that we might want to address:
* I recommend removing the direct execution of shell commands without displaying the command to the user first. In this scenario, the user would have the responsibility of reviewing and executing commands, including potentially destructive ones.
* It might be beneficial to add a loading indicator during sgpt requests. Some commands could take a few seconds to load, for instance, Using docker create postgress container, forward ports and create todo database. During the request, the shell appears to "hang" until the process is completed, which isn't the best experience for the user.
* There are instances where the first word of a natural language input exists as a command, leading to confusing results. For example, with find all files with json extension in current folder, find is a valid command for the shell. When we toggle, it provides a description for the shell command, even though this was meant as a natural language request. Another example is install docker ('install' is a valid shell command for Mac users). A potential solution for these cases could involve using different key bindings for transforming natural language inputs into shell commands, and vice versa. However, I suspect there may be more efficient solutions available.

chinarjoshi commented:
### Summary of changes

1. New feature: Explicitly translate buffer from natural language to shell. Example use case: "ls in long form" -> "ls -l". Default keybinding: Alt+T, and can be overridden with new config option `TOGGLE_NATURAL_LANGUAGE_KEYBINDING`.
2. Executing natural language now always prompts by default, and gives a clear warning message for destructive commands. Users can execute non-destructive commands without confirmation using a new config option `SHELL_EXECUTE_NO_CONFIRM`.
3. Executing natural language respects the existing config option `DEFAULT_EXECUTE_SHELL_CMD`.
5. Adds a loading indicator, ⌛ , whenever sgpt is invoked internally.
6. Default keybindings are now Alt+t for toggling buffer and Alt+e for explicitly executing natural language. Enter remains the same. These two keybindings can be overridden with config options `TOGGLE_BUFFER_KEYBINDING` and `EXECUTE_NATURAL_LANGUAGE_KEYBINDING` respectively.

chinarjoshi commented:
1. Changed default keybindings from Alt+* to Control+*.
2. Changed `zle redisplay` invocations to `zle -I && zle redisplay`. Hopefully this combination is compatible for both Mac and Linux. I was a bit confused that `zle -I` fixes because according to its man page it prepares the widget for stdout, which is strange that changing the buffer counts as stdout on Mac.
3. I think the simplest solution is to put the shell script in the main branch as `/scripts/.sgpt.sh` or `/scripts/integrations/*.sh`, so it gets automatically updated with pip. Then, the only command the user has to run to opt-in to the feature is `echo '. ~/.local/lib/python3.11/site-packages/sgpt/... >> ~/.zshrc`. This would require adding some line to pyproject.toml to include the script as a data file. I found a way to do this with [poetry](https://python-poetry.org/docs/pyproject/#include-and-exclude) and [setuptools](https://setuptools.pypa.io/en/latest/userguide/datafiles.html), but not without either.

TheR1D commented:
If we move integration scripts to main and distribute it with `sgpt` we will be forced to make new pip package updates every time we would need to update shell integration scripts. In my opinion shell integration scripts should be completely separate with independed update logic. I think we can't just `echo some.sh >> ~/.zshrc`, if user already has shell integration written in shell profile, it will just keep adding new version at the end of file. My idea was to find some way to update our shell integraion lines in user shell profiles. Also it would be nice to keep bash/zsh version separate and deliver only lines which are related to user `$SHELL`.

chinarjoshi commented:
Important change: I have moved the shell script to a separate repository to decouple it from the Python codebase and address the installation issue. This decoupling allows the script to be used with any command-line GPT program, while keeping shell_gpt as the default program. I also made a [pr to ohmyzsh](https://github.com/ohmyzsh/ohmyzsh/pull/11779)  to streamline installation.

https://github.com/chinarjoshi/englizsh/
